app/services/process_feed_service.rb

```diff
@@ -22,268 +22,285 @@ def process_entries(xml, account)
   class ProcessEntry
     def call(xml, account)
       @account = account
-      @xml     = xml
-
-      return if skip_unsupported_type?
-
-      case verb
-      when :post, :share
-        return create_status
-      when :delete
-        return delete_status
-      end
+      @fetched = Activity.new(xml)
+
+      return unless [:activity, :note, :comment].include?(@fetched.type)
+
+      klass = case @fetched.verb
+              when :post
+                PostActivity
+              when :share
+                ShareActivity
+              when :delete
+                DeletionActivity
+              else
+                return
+              end
+
+      @fetched = klass.new(xml, account)
+      @fetched.perform
     rescue ActiveRecord::RecordInvalid => e
       Rails.logger.debug "Nothing was saved for #{id} because: #{e}"
       nil
     end
 
-    private
-
-    def create_status
-      if redis.exists("delete_upon_arrival:#{@account.id}:#{id}")
-        Rails.logger.debug "Delete for status #{id} was queued, ignoring"
-        return
+    class Activity
```

Comment:
I wouldn't be against extracting all these inner classes into their own files, for example under app/services/ostatus/activities - your patch is a fantastic improvement in code organization, but the outer class has also become much bigger because of it, and harder to read.